Add structured logging support to custom log states

CodeBlanch opened the pull request:
This PR updates the LogMetric, LogFunctionResult, LogFunctionResultAggregate ILogger extensions to use TStates which support structured logging:

* `{OriginalFormat}` attribute is added as the last KVP which contains a template for these logs.
* A formatter is provided which will generate formatted messages based on the template if called.

The reason for this is OpenTelemetry is implemented as an ILoggerProvider and it kind of barfs on the existing states which do not implement the typical structured logging state contract. Implementing this structure will make these logs work better with OpenTelemetry and likely many other ILoggerProvider implementations. See: https://github.com/Azure/azure-sdk-for-net/issues/37832

This is similar to: https://github.com/dotnet/aspnetcore/pull/45253

CodeBlanch committed:
Add structured logging support to custom log states.

CodeBlanch committed:
Make sure formatters don't capture locals.

RohitRanjanMS commented:
This looks good.
<img width="1328" alt="image" src="https://github.com/Azure/azure-webjobs-sdk/assets/90008725/bf20f866-01fd-43aa-8dc7-996642871c6f">

I will check at other data sources and confirm.

RohitRanjanMS committed:
Update LoggerExtensions.cs

We ignore these `Aggregate`, `Result` and `Metric` logs and don't log it anywhere. We rely the on formatted message to ignore these logs (one example here https://github.com/Azure/azure-functions-host/blob/b42df4d4e63ab87f7a737f5a65dadf3eb7ba7de4/src/WebJobs.Script/Diagnostics/FileLogger.cs#L50). 

Setting the formatter to `null` so that we still ignore these.

RohitRanjanMS committed:
Update FunctionResultAggregateState.cs

Updated format string.

RohitRanjanMS committed:
Update FunctionResultState.cs

Updated formatting.

RohitRanjanMS committed:
Update LoggerExtensionsTests.cs

Fixing UTs

RohitRanjanMS commented:
Hi @CodeBlanch , I have made some changes based on our last conversation. Also updated formatting to align with rest of the logs. If you are ok with these changes, I will merge the code and create a release.

CodeBlanch commented:
@RohitRanjanMS Changes LGTM!

cataggar commented:
This got merged on August 21st, so it looks like it should be in 3.0.38 or higher, but none are yet published to https://www.nuget.org/packages/Microsoft.Azure.WebJobs/
